compiler/rustc_middle/src/ty/layout.rs

```diff
@@ -3482,3 +3482,143 @@ fn make_thin_self_ptr<'tcx>(
+///
+/// FIXME: Once all the conservatism is removed from here, and the checks are ran by default,
+/// we can use the const evaluation checks always instead.
+pub fn might_permit_raw_init<'tcx, C>(this: TyAndLayout<'tcx>, cx: &C, init_kind: InitKind) -> bool
+where
+    C: HasDataLayout + ty::layout::HasParamEnv<'tcx> + ty::layout::HasTyCtxt<'tcx>,
+{
+    return might_permit_raw_init_inner(this, cx, init_kind, false);
+
+    fn might_permit_raw_init_inner<'tcx, C>(
+        this: TyAndLayout<'tcx>,
+        cx: &C,
+        init_kind: InitKind,
+        inside_array: bool,
+    ) -> bool
+    where
+        C: HasDataLayout + ty::layout::HasParamEnv<'tcx> + ty::layout::HasTyCtxt<'tcx>,
+    {
+        let scalar_allows_raw_init = move |s: Scalar| -> bool {
+            match init_kind {
+                InitKind::Zero => {
+                    // The range must contain 0.
+                    s.valid_range(cx).contains(0)
+                }
+                InitKind::Uninit => {
+                    // FIXME(#66151) This should be "is the type a union", but that's pending on a
+                    // resolution to https://github.com/rust-lang/unsafe-code-guidelines/issues/71
+                    // And likely a large number of crates fail with those rules, though no crater
+                    // run has been done yet.
+                    //
+                    // The range must include all values.
+                    s.is_always_valid(cx)
+                }
+            }
+        };
+
+        // These are bypasses in order to try not to break quite as many crates so quickly.
+        // They are being done only if we're inside an array, to ensure we don't panic on *less*
+        // things than we did before this change.
+        // See: https://github.com/rust-lang/rust/pull/99389
+        if inside_array {
+            match init_kind {
+                // We panic if creating this type with all 0x01 bytes would
+                // cause LLVM UB.
+                //
+                // Therefore, in order for us to not panic,
+                // * the alignment of the pointer must be 1
+                //   (or we would have an unaligned pointer)
+                //
+                // * the statically known size of the pointee must be 0.
+                //   (or we would emit dereferenceable)
+                //
+                // If this bypass didn't exist, old versions of `hyper` with no semver compatible
+                // fix (0.11, 0.12, 0.13) would panic, as they make uninit &[u8] and &str.
+                InitKind::Uninit => {
+                    if let ty::Ref(_, inner, _) = this.ty.kind() {
+                        let penv = ty::ParamEnv::reveal_all().and(*inner);
+                        if let Ok(l) = cx.tcx().layout_of(penv) {
+                            return l.layout.align().abi == Align::ONE
+                                && l.layout.size() == Size::ZERO;
```

Comment:
Ah, I was too quick here... this does not work properly for `&dyn Trait`, we would still have LLVM UB there since the vtable is marked as 8-aligned and as dereferenceable for 24 bytes.

Our `FieldsShape::Arbitrary` branch below should catch that, but it means this early-return here may be wrong. So we probably do need to go back to what you had, where `[u8]` and `str` are special-cased, or we need to at least compute the type of the unsized tail and bail if it is a `dyn Trait`.

Please add a test ensuring that an uninit `[&dyn Send; 1]` correctly panics.

Reply:
Ah yeah, that did indeed incorrectly not panic there. I reverted it back to a `[T]` where align_of(T) == 1, or a `str` check.

Also added tests that `&dyn Trait` and `*const dyn Trait` both panic when uninitialized inside arrays.